### PR TITLE
Typed continuations: cont.bind instructions

### PR DESCRIPTION
**Pull request by frank-emrich**
This PR is part of a series that adds basic support for the [typed continuations/wasmfx proposal](https://github.com/wasmfx/specfx).

This particular PR adds support for the `cont.bind` instruction for partially applying continuations, documented [here](https://github.com/wasmfx/specfx/blob/main/proposals/continuations/Overview.md#instructions).

In short, these instructions are of the form `(cont.bind $ct_before $ct_after)` where `$ct_before` and `$ct_after` are related continuation types. They must only differ in the number of arguments, where `$ct_before` has _n_ additional parameters as compared to `$ct_after`, for some _n_ ≥ 0.
The idea is that `(cont.bind $ct_before $ct_after)` then takes a reference to a continuation of type `$ct_before` as well as _n_ operands and returns a (reference to a) continuation of type  `$ct_after`.
Thus, the folded textual representation looks like `(cont.bind $ct_before $ct_after arg1 ... argn c)`.

Support for the instruction is implemented in both the old and the new wat parser.

Note that this PR does not implement validation of the new instruction.

### CODE REVIEW

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -7768,6 +7772,37 @@ void WasmBinaryReader::visitRefAs(RefAs* curr, uint8_t code) {
...
+    }
+  }
+
+  curr->cont = popNonVoidExpression();
+
+  size_t paramsBefore =
+    curr->contTypeBefore.getContinuation().type.getSignature().params.size();
+  size_t paramsAfter =
+    curr->contTypeAfter.getContinuation().type.getSignature().params.size();
+  size_t numArgs = paramsBefore - paramsAfter;
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:7797]: It would be good to ensure that this doesn't underflow.

File: src/wasm/wasm-ir-builder.cpp [hunk 2]
```diff
@@ -619,6 +619,26 @@ Result<> IRBuilder::visitStringEncode(StringEncode* curr) {
...
+Result<> IRBuilder::visitContBind(ContBind* curr) {
+  auto cont = pop();
+  CHECK_ERR(cont);
+  curr->cont = *cont;
+
+  size_t paramsBefore =
+    curr->contTypeBefore.getContinuation().type.getSignature().params.size();
+  size_t paramsAfter =
+    curr->contTypeAfter.getContinuation().type.getSignature().params.size();
+  size_t numArgs = paramsBefore - paramsAfter;
```
Comment (tlively) [inline src/wasm/wasm-ir-builder.cpp:631]: Same here about the underflow.

**Review by tlively — APPROVED** [review]
LGTM with those underflow checks 👍

**Comment by frank-emrich** [comment]
Thanks! Should be all done now.

**Pull request merged**